### PR TITLE
drivers: nrf: Add power management to nrf TWIM peripheral.

### PR DESCRIPTION
**Pull request by Mierunski**
Add power management to nrf TWIM peripheral.

Signed-off-by: Mieszko Mierunski <mieszko.mierunski@nordicsemi.no>

### CODE REVIEW

**Review by anangl — APPROVED** [review]

**Comment by anangl** [comment]
@Mierunski Could you provide similar PRs for TWI and SPI peripherals?

**Review by anangl — APPROVED** [review]

File: drivers/i2c/i2c_nrfx_twim.c [hunk 1]
```diff
@@ -169,31 +205,32 @@ static int init_twim(struct device *dev, const nrfx_twim_config_t *config)
...
-			    POST_KERNEL,				       \
-			    CONFIG_I2C_INIT_PRIORITY,			       \
-			    &i2c_nrfx_twim_driver_api)
+	DEVICE_DEFINE(twim_##idx,					       \
+		      DT_NORDIC_NRF_I2C_I2C_##idx##_LABEL,		       \
+		      twim_##idx##_init,				       \
+		      twim_nrfx_pm_control,				       \
+		      &twim_##idx##_data,				       \
+		      &twim_##idx##z_config,				       \
+		      POST_KERNEL,					       \
```
Comment (ioannisg) [inline drivers/i2c/i2c_nrfx_twim.c]: Just wondering @Mierunski why is this in a POST_KERNEL initialization?

File: drivers/i2c/i2c_nrfx_twi.c [hunk 2]
```diff
@@ -148,6 +150,40 @@ static int init_twi(struct device *dev, const nrfx_twi_config_t *config)
...
+static int twi_nrfx_pm_control(struct device *dev, u32_t ctrl_command,
+				void *context, device_pm_cb cb, void *arg)
+{
+	static u32_t current_state = DEVICE_PM_ACTIVE_STATE;
+
+	if (ctrl_command == DEVICE_PM_SET_POWER_STATE) {
+		u32_t new_state = *((const u32_t *)context);
+
+		if (new_state != current_state) {
+			if (new_state == DEVICE_PM_ACTIVE_STATE) {
```
Comment (pizi-nordic) [inline drivers/i2c/i2c_nrfx_twi.c]: Could we have switch/case here? Please note that "if" requires priority encoding which is not needed for switch case so the resulting code is smaller and faster.

File: drivers/i2c/i2c_nrfx_twim.c [hunk 3]
```diff
@@ -137,20 +138,55 @@ static const struct i2c_driver_api i2c_nrfx_twim_driver_api = {
...
+	static u32_t current_state = DEVICE_PM_ACTIVE_STATE;
+
+	if (ctrl_command == DEVICE_PM_SET_POWER_STATE) {
+		u32_t new_state = *((const u32_t *)context);
+
+		if (new_state != current_state) {
+			if (new_state == DEVICE_PM_ACTIVE_STATE) {
+				init_twim(dev);
+			} else {
+				assert(new_state == DEVICE_PM_LOW_POWER_STATE ||
```
Comment (pizi-nordic) [inline drivers/i2c/i2c_nrfx_twim.c]: Could we have switch/case here? Please note that "if" requires priority encoding which is not needed for switch case so the resulting code is smaller and faster.

Comment (pizi-nordic) [inline drivers/i2c/i2c_nrfx_twim.c]: It looks that we only support FORCE_SUSPEND state. Is this what we want?

File: drivers/i2c/i2c_nrfx_twim.c [hunk 4]
```diff
@@ -137,20 +138,55 @@ static const struct i2c_driver_api i2c_nrfx_twim_driver_api = {
...
+			dev->config->name);
 		return -EBUSY;
 	}
 
+
 	return 0;
 }
 
+#ifdef CONFIG_DEVICE_POWER_MANAGEMENT
+static int twim_nrfx_pm_control(struct device *dev, u32_t ctrl_command,
```
Comment (pizi-nordic) [inline drivers/i2c/i2c_nrfx_twim.c]: It looks that runtime configuration (i2c_api_configure_t) is not preserved when power mode is changed.

**Review by pizi-nordic — CHANGES_REQUESTED** [review]

**Comment by Mierunski** [comment]
@carlescufi @pizi-nordic ping

**Review by pizi-nordic — APPROVED** [review]

File: drivers/i2c/i2c_nrfx_twim.c (hunk 1 above)
Comment (ioannisg) [reply drivers/i2c/i2c_nrfx_twim.c]: Ping @Mierunski

**Review by carlescufi — APPROVED** [review]

**Pull request merged**